(GH-61) Fix choco install package result

esiebes opened the pull request:
edit comments from pullrequest 102



Supersedes #102

Related to #61

bassedball committed:
add Invoke-Chocolatey cmdlet which throws error on choco failure

bassedball committed:
Merge branch 'development' into feature/chocoinstallResult

bassedball committed:
removed typo

esiebes committed:
removed alias, ensured LASTEXITCODE powershell is set same as choco exitcode. Some casting

esiebes commented:
new pullrequest originally pullrequest 102.

esiebes committed:
add lastexitcode after invoke-chocolatey, included valid exit codes

esiebes committed:
some fixes

esiebes committed:
removed unused variable

esiebes committed:
invoke-chocolatey now returns object array (same as chocolatey

esiebes committed:
inovke chocolatey now returning output

esiebes committed:
change to output

pauby commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2017 Chocolatey Software, Inc.
```
Can you remove this change from the commit?

pauby commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -240,13 +240,9 @@ function InstallPackage
     Write-Verbose -Message "Install command: 'choco install $pName $chocoinstallparams'"
-
-    $packageInstallOuput = Invoke-Expression -Command "choco install $pName $chocoinstallparams"
+    $packageInstallOuput = Invoke-ChocoLatey "install $pName $chocoinstallparams"
```
Picky but can you change Invoke-ChocoLatey to a lowercase 'l'?
Can you also add the arguments parameter explicitly?

pauby commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -240,13 +240,9 @@ function InstallPackage
     Write-Verbose -Message "Package output $packageInstallOuput "
 
-    # Clear Package Cache
-    Get-ChocoInstalledPackage 'Purge'
```
Is there a reason this has been removed?

esiebes replied:
the purge command can cause errors and hangs code in DSC configurations.

pauby replied:
There are no issues submitted around the purge command. If this isn't the fix for the issue itself can you raise an issue we can look at and remove it from this PR?

pauby commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -266,19 +262,16 @@ function UninstallPackage
     Write-Verbose -Message "Package uninstall output $packageUninstallOuput "
 
-    # Clear Package Cache
-    Get-ChocoInstalledPackage 'Purge'
```
Same here?

esiebes replied:
he purge command can cause errors and hangs code in DSC configurations.

pauby replied:
See previous comment.

esiebes replied:
I added the purge command again.

pauby commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -389,41 +383,81 @@ Function Upgrade-Package {
     }
-    $cmd = "choco upgrade $pName $chocoupgradeparams"
+    $cmd = "upgrade $pName $chocoupgradeparams"
     Write-Verbose -Message "Upgrade command: '$cmd'"
```
I'd like to keep that Verbose message consistent from what it was previously - prepend it with 'choco'?

esiebes replied:
done

pauby replied:
I don't see this change in the latest commit.

esiebes replied:
Removed the choco

pauby commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -389,41 +383,81 @@ Function Upgrade-Package {
-    Get-ChocoInstalledPackage 'Purge'
+function Get-ChocoInstalledPackage {
+    Return (choco list -lo -r | ConvertFrom-Csv -Header 'Name', 'Version' -Delimiter "|")
 }
```
What's your thinking behind running this every time rather than caching it for 60 seconds?
The Return isn't needed.

esiebes replied:
In some instances of DSC configuratioin 60 seconds is to long if a small package is installed, so the test will fail I will remove the Return

pauby replied:
See previous comment around this.

pauby commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -389,41 +383,81 @@ Function Upgrade-Package {
+    #Set $LASTEXITCODE variable.
+    powershell.exe   -NoLogo -NoProfile -Noninteractive "exit $exitcode"
+
+    if($exitcode -in $validExitCodes )
```
What happens if a non-MSI exits with one of the MSI exit codes? That could indicate an error but it's ignored?

esiebes replied:
its ignore/ non error

pauby replied:
If an EXE exits with an error code that matches one of those we need that to come through. Ignoring it would indicate it installed it okay when in fact it didn't.

pauby commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -389,41 +383,81 @@ Function Upgrade-Package {
+    $p.Start() | Out-Null
+
+    $output = $p.StandardOutput.ReadToEnd()
+    $p.WaitForExit()
```
By default choco.exe will timeout after 45 minutes. It can be changed in the config but I'm wondering if we should kill it before then. Thoughts?

esiebes replied:
45 minutes is a long time, but i can imagin that in some instances an install of a package takes longer than that. For instance a full install of visual studio on a slow internet connection.

pauby commented:
@esiebes Is this still a PR you are wanting to work on?

esiebes commented:
yes, i will work on it asap

Op zo 28 jul. 2019 16:42 schreef Paul Broadwith <notifications@github.com>:

> @esiebes <https://github.com/esiebes> Is this still a PR you are wanting
> to work on?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chocolatey/cChoco/pull/103?email_source=notifications&email_token=AIEO72WOUTKY5QGHITNINVDQBWV4DA5CNFSM4EIE6TEKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD27AA2I#issuecomment-515768425>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIEO72WVMNTQ2JGIIONXJM3QBWV4DANCNFSM4EIE6TEA>
> .
>

pauby commented:
@esiebes Are you still working on this?

esiebes committed:
some minor symantic changes

pauby commented on DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1:
```diff
@@ -325,7 +324,7 @@ Function Test-LatestVersionInstalled {
     Write-Verbose -Message "Testing if $pName can be upgraded: 'choco upgrade $pName $chocoupgradeparams'"
 
-    $packageUpgradeOuput = Invoke-Chocolatey "upgrade $pName $chocoupgradeparams"
+    $packageUpgradeOuput = Invoke-Chocolatey "choco upgrade $pName $chocoupgradeparams"
```
The `choco` command shouldn't be passed to Invoke-Chocolatey.

esiebes committed:
reverted some changes

pauby commented:
@esiebes Are you still wanting to take this forward?

esiebes commented:
Yes, will forward the changes tomorrow.

Op do 21 mei 2020 18:01 schreef Paul Broadwith <notifications@github.com>:

> @esiebes <https://github.com/esiebes> Are you still wanting to take this
> forward?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chocolatey/cChoco/pull/103#issuecomment-632176997>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIEO72RTB2W56QDGKCM4DGTRSVF45ANCNFSM4EIE6TEA>
> .
>

esiebes committed:
mergeconflicts

esiebes committed:
typo

esiebes committed:
Created function invoke-chocolatey to ensure powershell exitcode equals
choco exitcode
typo

missed some bracket

add missing function

else block added, gone after merge

 should process with test not used

 add shouldprocess to Invoke chocolatey

 add verbose to scriptstest

 shouldprocess thing error

no error local found with shouldprocess

mm

esiebes committed:
missed some bracket

esiebes committed:
add missing function

esiebes committed:
else block added, gone after merge

esiebes committed:
should process with test not used

esiebes committed:
add shouldprocess to Invoke chocolatey

esiebes committed:
add verbose to scriptstest

esiebes committed:
shouldprocess thing error

esiebes committed:
no error local found with shouldprocess

esiebes committed:
mm

esiebes commented:
I keep getting a failed test in appveyor on the psscriptanalyzer test. Local on my machine i dont get it. Someon an idea?
[-] passes the PSScriptAnalyzer Rule PSShouldProcess 986ms
        Expected 0, but got 1.
        [-] passes the PSScriptAnalyzer Rule PSShouldProcess 986ms
        Expected 0, but got 1.
        34:             (Invoke-ScriptAnalyzer -Path $module.FullName -IncludeRule $rule.RuleName ).Count | Should Be 0
        at <ScriptBlock>, C:\projects\cchoco\tests\cChoco_ScriptAnalyzerTests.ps1: line 3434:             (Invoke-ScriptAnalyzer -Path $module.FullName -IncludeRule $rule.RuleName ).Count | Should Be 0
        at <ScriptBlock>, C:\projects\cchoco\tests\cChoco_ScriptAnalyzerTests.ps1: line 34

pauby commented:
Are you running the same version of PSScriptAnalzyer?

Can I also ask you to update some of those commit messages and preferably squash some of them?

esiebes commented:
I will do that. Whih versiin is used?

Op di 23 jun. 2020 13:28 schreef Paul Broadwith <notifications@github.com>:

> Are you running the same version of PSScriptAnalzyer?
>
> Can I also ask you to update some of those commit messages and preferably
> squash some of them?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chocolatey/cChoco/pull/103#issuecomment-648086091>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIEO72VIAGJFDTYI35357TLRYCGWBANCNFSM4EIE6TEA>
> .
>

pauby commented:
See [this](https://github.com/chocolatey/cChoco/pull/145#discussion_r444163663) for the maximum version that the author believes is supported.

Also see this issue #147 for the build issues.

esiebes committed:
Merge branch 'feature/chocoinstallresult' of https://github.com/esiebes/cChoco into feature/chocoinstallresult
